src/renderer/renderer.js

```diff
@@ -1,316 +1,639 @@
-import { toRad } from "../utils/math.js";
-import { getGL, glErrnoToMsg } from "../utils/gl_utils.js";
+import { isAnyArray } from "../utils/type_check.js";
+import { getGLContext } from "../utils/gl_utils.js";
 import { LoaderMerge } from "../utils/loader.js";
 import { GameComponent } from "../game_component.js";
 
-import { Buffers } from "./buffers.js";
+import { Buffer } from "./buffers.js";
+import { makeUniformsObj } from "./uniforms.js";
+import { ShaderProgram } from "./shader_program.js";
+import { Camera } from "./gl_camera.js";
 import { AtlasLoader } from "./atlas_data.js";
-import { ShaderLoader } from "./shader_loader.js";
-import { CubeDataAdder } from "./face_culling.js";
+import { ShaderProgramLoader } from "./shader_loader.js";
 import { ElementBundler } from "./vertex_bundle.js";
+import { assert } from "../utils/assert.js";
 
 /**
+ * @typedef {import('../game.js').Game} Game
  * @typedef {import('../world/chunk.js').Chunk} Chunk
  * @typedef {import('./chunk_renderer.js').ChunkRenderer} ChunkRenderer
  */
 
 // NOTE:
-// West  = +x
-// Up    = +y
 // North = +z
+// Up    = +y
+// East  = +z
 
 // thing i could eventually read:
 // https://www.toptal.com/game/video-game-physics-part-i-an-introduction-to-rigid-body-dynamics
 
 // TODO: switch to typescript??
-export class Renderer extends GameComponent {
+export class RenderMgr extends GameComponent {
+  /** @type {WebGLRenderingContext} */
+  gl;
+  /** @type {PickingIdRenderer} */
+  pickingRenderer;
```

Comment:
Typescript keeps complaining about these 2:
```
'gl' is defined as an accessor in class 'GameComponent', but is overridden here in 'RenderMgr' as an instance property
```
TODO: make them properties that just access a 'protected' `_gl` attribute. (Typescript wants these to be properties because they're properties on the `GameComponent` class). Or make them not `GameComponent` but just have `.game` attribute.

Reply:
If option 2, need to add a `GameComponentMixin` class so that subclasses can use `GameComponent` if they want and aren't locked out of it.